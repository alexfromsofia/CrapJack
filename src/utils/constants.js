export const CARD_POINTS = {
  2: 2,
  3: 3,
  4: 4,
  5: 5,
  6: 6,
  7: 7,
  8: 8,
  9: 9,
  10: 10,
  JACK: 10,
  QUEEN: 10,
  KING: 10,
  ACE: 10
};

export const REQUEST_METHODS = {
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT',
  DELETE: 'DELETE'
};

export const SYSTEM_MESSAGES = {
  WELCOME: 'Welcome to Crap Jack!'
};
