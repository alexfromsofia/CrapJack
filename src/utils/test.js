import React from 'react';

import { StoreProvider } from '../store/StoreProvider';
import { initialState } from '../store/gameReducer';
import { rootReducer, init } from '../store';

export const withProvider = (Component, props = {}) => (
  <StoreProvider initialState={initialState} reducer={rootReducer} init={init}>
    <Component {...props} />
  </StoreProvider>
);
