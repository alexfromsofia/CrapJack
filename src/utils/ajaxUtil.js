import { REQUEST_METHODS } from '../utils/constants';

/**
 * @param {String} method either GET,PUT,POST,DELETE
 * @param {String} endpoint the url from which to fetch from
 */
export const generateRequest = async ({
  method = REQUEST_METHODS.GET,
  endpoint = ''
}) => {
  try {
    const response = await fetch(
      // Hardcoding for the purpose of the task. Usually will be a environment variable or an argument.
      `https://deckofcardsapi.com/api/deck/${endpoint}`,
      {
        method
      }
    );

    const { status, ok } = response;

    if (ok) {
      return response.json();
    }

    if (status >= 400) {
      console.log(response);
    }
  } catch (error) {
    console.error(error);

    return error;
  }
};

export const getNewShuffledDeck = async () => {
  try {
    const deck = await generateRequest({
      endpoint: 'new/shuffle/?deck_count=1'
    });

    if (!deck.success) {
      return { error: 'Error getting deck.' };
    }

    return deck;
  } catch (error) {
    console.log(error);
  }
};

export const drawCards = async (deckId = null, numOfCards = 1) => {
  if (!deckId) {
    return console.log('Please provide deckId');
  }

  try {
    const cards = await generateRequest({
      endpoint: `${deckId}/draw/?count=${numOfCards}`
    });

    if (!cards.success) {
      return { error: 'Error getting cards.' };
    }

    return cards;
  } catch (error) {
    console.log(error);
  }
};
