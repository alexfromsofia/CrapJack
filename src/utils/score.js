import { CARD_POINTS } from './constants';

export const calculateScore = (cards) =>
  cards.reduce((acc, card) => {
    const points = CARD_POINTS[card.value];

    acc += points;

    return acc;
  }, 0);
