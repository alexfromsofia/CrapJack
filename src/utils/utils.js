import { CARD_POINTS, BUST_TRESHOLD, PLAYER_TYPES } from './constants';

/**
 * Calculations are made based on the rules.
 * 2 = 2pts, 3 = 3pts,...10 = 10pts, J, Q, K, A = 10pts each
 * @param {Array} cards
 * @returns {Number}
 */
export const calculateScore = (cards) => {
  if (!Array.isArray(cards)) return 0;

  return cards.reduce((acc, card) => {
    const points = CARD_POINTS[card.value] || 0;

    acc += points;

    return acc;
  }, 0);
};

/**
 * Returns a string based on the game conditions for winning.
 * Returns null if there is no winner.
 * @param {Object} dealer
 * @param {Object} player
 * @returns {String or null}
 */
export const determineWinner = ({ dealer, player }) => {
  const { score: dealerScore } = dealer;
  const { score: playerScore } = player;

  // No winner if both scores are above 21 or equal to each other.
  const areBothScoresBust =
    (dealerScore > BUST_TRESHOLD && playerScore > BUST_TRESHOLD) ||
    dealerScore === playerScore;

  if (areBothScoresBust) return null;

  if (dealerScore > BUST_TRESHOLD) {
    return PLAYER_TYPES.PLAYER;
  }

  if (playerScore > BUST_TRESHOLD) {
    return PLAYER_TYPES.DEALER;
  }

  return playerScore > dealerScore ? PLAYER_TYPES.PLAYER : PLAYER_TYPES.DEALER;
};

/**
 * Returns a random number between the specified values.
 * The returned value is no lower than (and may possibly equal) min, and is less than (and not equal) max.
 * @param {Number} min
 * @param {Number} max
 * @returns {Number}
 */
export const getRandomArbitrary = (min, max) =>
  Math.random() * (max - min) + min;
