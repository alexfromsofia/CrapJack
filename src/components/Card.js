import React, { useState } from 'react';
import classNames from 'classnames';
import { getRandomArbitrary } from '../utils/utils';
import { useStoreValue } from '../store/StoreProvider';
import CardFront from '../img/card_front.svg';
import { PLAYER_TYPES } from '../utils/constants';

function Card({ card, playerType }) {
  const [loaded, setLoaded] = useState(false);
  const [{ game }] = useStoreValue();
  const { isRevealed } = game;
  const handleCardLoad = () => {
    setTimeout(() => {
      setLoaded(true);
      // Delay card animation to simulate separate card dealing.
    }, getRandomArbitrary(200, 500));
  };
  const isFlipped = playerType === PLAYER_TYPES.PLAYER || isRevealed;

  return (
    <div
      className={classNames('scene', playerType, {
        loaded
      })}
    >
      <div className={classNames('game__card', { 'is-flipped': isFlipped })}>
        <div className="game__card--face front">
          <img src={CardFront} />
        </div>
        <div className="game__card--face back">
          <img src={card.images.png} alt={card.code} onLoad={handleCardLoad} />
        </div>
      </div>
    </div>
  );
}

export default Card;
