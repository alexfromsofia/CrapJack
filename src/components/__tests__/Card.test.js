import { shallow } from 'enzyme';
import Card from '../Card';
import { withProvider } from '../../utils/test';

const props = { code: '1', images: { png: '' } };
let component;

beforeEach(() => {
  component = shallow(withProvider(Card, props)).find(Card);
});

it('Matches snapshot', () => {
  expect(component).toMatchSnapshot();
});

it('To match props', () => {
  expect(component.props()).toEqual(props);
});
