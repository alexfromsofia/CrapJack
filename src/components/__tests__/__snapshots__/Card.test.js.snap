// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<Card
  code="1"
  images={
    Object {
      "png": "",
    }
  }
/>
`;
